### ISSUE TITLE
Same migration running twice...when `transaction: false`

### ISSUE THREAD

**Issue by newhouse** [knex/knex#4495]
# Environment

Knex version: 0.21.12
Database + version: Postgres 12.6
OS: Linux

# Bug

We use K8S and leverage `initContainer`s to be responsible for running our migrations via `knex migrate:latest` when we deploy, before the pods running the new code are fired up.

If we have more than 1 pod, that means more than 1 `initContainer` will run. It is expected that 1 init container will be successful at running any outstanding migration(s), while the other(s) will either finish with an `already up to date`, or warn about obtaining the lock, and then fail/retry and subsequently complete with `already up to date`.

However, I have recently deployed code that contains a migration that was actually run twice: once by each init container (we have 2 pods, so 2 init containers).
The init container "A" did the work, while init container "B" failed at first complaining about the lock...but then subsequently re-ran the migration.

2 entries were created in the `knex_migrations` table as a result - each with the same value in the `name` column, but with different `batch` numbers and timestamps.

I have been been able to replicate and observe this behavior multiple times, though it does not always happen...presumably due to timing.

Some notable things:
1. This was a "data" migration where we did not actually change the schema, but wanted to convert some data.
2. Because there was a lot of data being altered, the migration exported a `config` that specified not to use a transaction.

I'm confused by what Knex must be doing with the locks, as well as by its decision to re-run the migration with the same filename...seems like a bug around locking and selecting/inserting into the `knex_migration` table both when it's thinking about doing the work, and when it has completed doing the work.

Init container A's logs were something like this:
```sh
knex migrate:latest --cwd server
Batch 76 run: 1 migrations
```

While init container B's logs were something like this:
```sh
knex migrate:latest --cwd server
Can't take lock to run migrations: Migration table is already locked
If you are sure migrations are not running you can release the lock manually by running 'knex migrate:unlock'
error Command failed with exit code 1.
knex migrate:latest --cwd server
Batch 77 run: 1 migrations
```

Here's the `knex_migrations` table afterwards:
![image](https://user-images.githubusercontent.com/2524009/119152560-45961a80-ba05-11eb-84a4-274d2aa75458.png)

Here's a simplified version of the migration file:
```js
'use strict'

export const config = {
  transaction: false,
}

export async function up (knex) {
  while (true) {
    // Select some things
    if (lengthOfThings === 0) break
    // Do some work    
  }
  
  return true
}

export function down (knex) {
  return Promise.resolve()
}
```

**Comment by KristjanTammekivi** [comment]
Do you have readinessprobes on the pods?

**Comment by KristjanTammekivi** [comment]
What I was thinking was that maybe the first pod became ready before the migrations are done, but no, initcontainer should be finished before that happens, right?

**Comment by newhouse** [comment]
I believe that that the `initContainer`s run and must exit successfully **_before_** the nodes in the Pod will be started up. But, as I mentioned, we have 2 Pods so each has an `initContainer` that runs at roughly the same time.

**Comment by ungrim97** [comment]
@newhouse Have you tried this against the latest version of KnexJS (0.95.5)?

**Comment by mercmobily** [comment]
This is a duplicate of #6301. This use case is actually covered by that issue.
